lib/bolt/cli.rb

```diff
@@ -562,6 +571,12 @@ def apply_manifest(code, targets, filename = nil, noop = false)
       results.ok ? 0 : 1
     end
 
+    def do_nothing_manifest?(ast)
+      ast.body.is_a?(Puppet::Pops::Model::Definition) ||
```

Comment:
Hm, this seems a bit flawed. It doesn't catch some basic cases, like a completely empty manifest file or a class with no resources, while it *does* warn for some manifests that do define resources, such as a manifest that has node definitions:
```
# test.pp
node default {
file { '/root/foo.txt': ensure => present }
}
```
I think the main point of this ticket is to warn people who have neglected to declare their classes, which it does, in which case we may want to refine this function to recursively inspect NodeDefinitions and to include `Puppet::Pops::Model::Nop` as a check, though it's hard to be sure those are the only false cases.

Another approach would be to look for resource statements, either by stringifying the ast or walking it, neither of which is appealing...

Reply:
I was thinking the purpose of the ticket is to warn people who are trying to apply a class via `bolt apply <modules>/manifests/<class>.pp` rather than by declaring it.

I don't think we should warn in the `NodeDefinition` case as node definitions are declared implicitly. If we're not warning there, then I also don't think we need to worry about the `BlockExpression` case as ordinary class/define manifests are not allowed to have more than a single definition (they may technically be allowed to have typedefs, but I wouldn't worry about that).

So that would make this simply `if ast.body.is_a?(Puppet::Pops::Model::HostClassDefinition) || ast.body.is_a?(Puppet::Pops::Model::ResourceTypeDefinition)`

Reply:
I guess to put it another way, my thought for this ticket was that we want to identify a _usage_ error (you invoked `bolt apply` with what is effectively a bad argument) vs. a _logical_ error (you ran an apply that generated no resources).